lib/content-services/src/lib/tree/components/tree.component.ts

```diff
@@ -0,0 +1,250 @@
+        this.treeService.getSubNodes('-root-', skipCount, maxItems).subscribe((response: TreeResponse<T>) => {
+            this.treeService.treeNodes = response.entries;
+            this.treeNodesSelection.deselect(...response.entries);
+            this.paginationChanged.emit(response.pagination);
+            this.loadingRootSource.next(false);
+        });
+    }
+
+    /**
+     * Collapses or expanding the node based on its current state
+     *
+     * @param node node to be collapsed/expanded
+     */
+    public expandCollapseNode(node: T): void {
+        if (this.treeService.treeControl.isExpanded(node)) {
+            this.treeService.collapseNode(node);
+        } else {
+            node.isLoading = true;
+            this.treeService.getSubNodes(node.id, 0, this.userPreferenceService.paginationSize).subscribe((response: TreeResponse<T>) => {
+                this.treeService.expandNode(node, response.entries);
+                this.paginationChanged.emit(response.pagination);
+                node.isLoading = false;
+                if (this.treeNodesSelection.isSelected(node)) {
+                    //timeout used to update nodeCheckboxes query list after new nodes are added so they can be selected
+                    setTimeout(() => {
+                        this.treeNodesSelection.select(...response.entries);
+                    });
+                }
+            });
+        }
+    }
+
+    /**
+     * Loads more subnode for a given parent node
+     *
+     * @param node parent node
+     */
+    public loadMoreSubnodes(node: T): void {
+        node.isLoading = true;
+        const parentNode: T = this.treeService.getParentNode(node.parentId);
+        this.treeService.treeNodes.splice(this.treeService.treeNodes.indexOf(node), 1);
+        const loadedChildren: number = this.treeService.treeNodes.filter((treeNode: T) => treeNode.parentId === parentNode.id).length;
+        this.treeService.getSubNodes(parentNode.id, loadedChildren, this.userPreferenceService.paginationSize).subscribe((response: TreeResponse<T>) => {
+            this.treeService.appendNodes(parentNode, response.entries);
+            this.paginationChanged.emit(response.pagination);
+            node.isLoading = false;
+            if (this.treeNodesSelection.isSelected(parentNode)) {
+                //timeout used to update nodeCheckboxes query list after new nodes are added so they can be selected
+                setTimeout(() => {
+                    this.treeNodesSelection.select(...response.entries);
+                });
+            }
+        });
+    }
+
+    /**
+     * When node is selected it selects all its descendants
+     *
+     * @param node selected node
+     */
+    public onNodeSelected(node: T): void {
+        this.treeNodesSelection.toggle(node);
+        const descendants: T[] = this.treeService.treeControl.getDescendants(node).filter((descendant: T) => descendant.nodeType !== TreeNodeType.LoadMoreNode);
```

Comment:
there's a lot of similar calls to the filter like this: `(descendant: T) => descendant.nodeType !== TreeNodeType.LoadMoreNode`

maybe worth moving filter to a separate function and just referencing the function... in this case the code will be cleaner and more readable.

see sort/reduce helpers example: https://github.com/Alfresco/alfresco-content-app/blob/develop/projects/aca-shared/src/lib/services/app.extension.service.ts#L386

Reply:
Moved to separate function